### PR TITLE
feat(components): add file uploader

### PR DESCRIPTION
**Pull request by nyan07**
## Description of the change

- Adds a FileUploader component to display the status of an upload
- Adds the FileUploaderButton, which could be used together with the FileUploader
- Adds a Uppy story example
- 

image

image

## Testing the change

- Run storybook

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Non-Breaking Change (change to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [x] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached.

### CODE REVIEW

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  fileSize?: string;
+
+  /** An error message */
+  errorMessage?: string;
+
+  /** The upload progress */
+  progress?: number;
+
+  /** The call to action button */
+  children: ReactElement<ButtonProps | FileUploaderButtonProps>;
```
Comment (apvale) [inline packages/components/src/components/FileUploader/FileUploader.tsx:38]: I know that I'm always asking this because I thought we wanted mandatory fields before the optional ones. But I don't know what the advantage of having it as well is 😅 What do you think?
My only point is that we can have it organized.

File: packages/components/src/components/FileUploader/FileUploader.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  fileName: "file_to_upload_with_a_long_name_v2.pdf",
+  progress: 50,
+  children: (
+    <FileUploaderButton trailingIcon="ArrowUpTrayIcon" variant="secondary">
+      Upload
+    </FileUploaderButton>
+  ),
+};
+
+export const Success = Template.bind({});
```
Comment (apvale) [inline packages/components/src/components/FileUploader/FileUploader.stories.tsx:39]: I'm not sure, but shouldn't we render it as mobile here or maybe create a test for it? 
From what I saw, we are not testing mobile designers, right? I also don't know if this is possible with Chromatic.

Comment (apvale) [reply packages/components/src/components/FileUploader/FileUploader.stories.tsx:39]: Yes, we can 👯 https://www.chromatic.com/docs/viewports
```
StoryName.parameters = {
  // Set the viewports in Chromatic at a story level.
  chromatic: { viewports: [320, 1200] },
};
```

File: packages/components/src/components/FileUploader/FileUploader.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    </FileUploaderButton>
+  ),
+};
+
+WithError.parameters = {
+  chromatic: { viewports: [320, 1200] },
+};
+
+export const WithUppy = (): React.ReactElement => {
+  const uppy = new Uppy();
```
Comment (rssilva) [inline packages/components/src/components/FileUploader/FileUploader.stories.tsx:97]: As much as I appreciate this effort a ton I wouldn't add uppy as a dependency on Pluto. Is that intended to stay?

Comment (rssilva) [reply packages/components/src/components/FileUploader/FileUploader.stories.tsx:97]: sure! It'll still be installed when storybook is deployed, right? Anyway lgtm

**Review by rssilva — APPROVED** [review]
📁 💯

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+  return (
+    <Button
+      {...props}
+      aria-label="Cancel upload"
+      iconOnly
+      trailingIcon="XMarkIcon"
+      variant="ghost"
+    />
+  );
+};
```
Comment (richbachman) [inline packages/components/src/components/FileUploader/FileUploader.tsx:117]: This file is becoming a bit large. Let's move this to a separate file.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+  return (
+    <Button
+      aria-label="Remove file"
+      iconOnly
+      trailingIcon="TrashIcon"
+      variant="ghost"
+    />
+  );
+};
```
Comment (richbachman) [inline packages/components/src/components/FileUploader/FileUploader.tsx:138]: Same for this. Separate file please.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+const getFileTitle = (label: string, fileUrl?: string): React.ReactElement => {
+  if (fileUrl) {
+    return (
+      <Anchor href={fileUrl} target="_blank">
+        {label}
+      </Anchor>
+    );
+  }
+  return <Text.span color="colorTextStrongest">{label}</Text.span>;
+};
```
Comment (richbachman) [inline packages/components/src/components/FileUploader/FileUploader.tsx:149]: Separate file. These are also components so should be named that way. Something like `FileUploaderTitle`.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+      onCancel,
+      onRemove,
+    },
+    ref
+  ) => {
+    const isMobile = useDown("md");
+    const status = getStatus(progress, fileUrl, errorMessage);
+
+    return (
+      <Box.div display="flex" flexDirection="column" gap="space25">
```
Comment (richbachman) [inline packages/components/src/components/FileUploader/FileUploader.tsx:199]: Can these Box elements be made separate components like `FileUploaderHeader` or `FileUploaderContent`. Something like that?

**Review by richbachman — CHANGES_REQUESTED** [review]
I feel like the main FileUploader file should be split up into more digestible pieces. There's a lot going on there, and its hard to parse it all.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+  if (!isEmpty(errorMessage)) return "error";
+  return "waiting";
+};
+
+const CancelUploadButton = (props: Partial<ButtonProps>): ReactElement => {
+  const isMobile = useDown("md");
+
+  if (isMobile) {
+    return (
+      <Button {...props} variant="secondary">
```
Comment (nyan07) [inline packages/components/src/components/FileUploader/FileUploader.tsx:102]: We need to add the `onRemove` prop here.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+        Cancel upload
+      </Button>
+    );
+  }
+
+  return (
+    <Button
+      {...props}
```
Comment (nyan07) [inline packages/components/src/components/FileUploader/FileUploader.tsx:110]: We need to add the `onRemove` prop here.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+    />
+  );
+};
+
+const RemoveButton = (props: Partial<ButtonProps>): ReactElement => {
+  const isMobile = useDown("md");
+
+  if (isMobile) {
+    return (
+      <Button {...props} aria-label="Remove file" variant="secondary">
```
Comment (nyan07) [inline packages/components/src/components/FileUploader/FileUploader.tsx:124]: We need to add the `onRemove` prop here.

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+        Remove
+      </Button>
+    );
+  }
+
+  return (
+    <Button
```
Comment (nyan07) [inline packages/components/src/components/FileUploader/FileUploader.tsx:131]: We need to add the `onRemove` prop here.

File: packages/components/src/components/FileUploader/FileUploader.tsx (hunk 10 above)
Comment (jvaladas) [reply packages/components/src/components/FileUploader/FileUploader.tsx:124]: Is this necessary here? We already have the onRemove on the parent component https://github.com/Localitos/pluto/blob/4afcf99ab45deb2bed2feddd09b01e137c45ea1b/packages/components/src/components/FileUploader/FileUploader.tsx#L265-L267

File: packages/components/src/components/FileUploader/FileUploader.tsx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      children,
+      progress = 0,
+      onCancel,
+      onRemove,
+    },
+    ref
+  ) => {
+    const isMobile = useDown("md");
+    const status = getStatus(progress, fileUrl, errorMessage);
+    const isMobileUploading = isMobile && status === "loading";
```
Comment (jvaladas) [inline packages/components/src/components/FileUploader/FileUploader.tsx:67]: This condition allows us to more clearly know when to render the mobile version of the upload since the structure is slightly different than all the other versions.

Mobile upload:
Screenshot 2022-11-09 at 11 08 26

Other mobile states:
Screenshot 2022-11-09 at 12 00 02

File: packages/components/src/components/FileUploader/FileUploaderIcon.tsx [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+}: FileUploaderIconProps): React.ReactElement => {
+  return (
+    <Box.div
+      alignItems="center"
+      borderRadius="borderRadius10"
+      display="flex"
+      h="36px"
+      justifyContent="center"
+      minWidth="36px"
+      w="36px"
```
Comment (apvale) [inline packages/components/src/components/FileUploader/FileUploaderIcon.tsx:66]: Shouldn't the minWidth be enough?

File: packages/components/src/components/FileUploader/FileUploaderProgressBar.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  const isMobile = useDown("md");
+
+  if (isMobile) {
+    return (
+      <Box.div paddingBottom={"space50"}>
+        <Box.span
+          color={"colorTextHeading"}
+          display={"block"}
+          fontSize={"fontSize10"}
+          paddingBottom={"space30"}
```
Comment (apvale) [inline packages/components/src/components/FileUploader/FileUploaderProgressBar.tsx:24]: 💅🏻 
```suggestion
          color="colorTextHeading"
          display="block"
          fontSize="fontSize10"
          paddingBottom="space30"
```

**Review by apvale — APPROVED** [review]

File: packages/components/src/components/FileUploader/types/fileUploaderStatus.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+export type FileUploaderStatus =
```
Comment (nyan07) [inline packages/components/src/components/FileUploader/types/fileUploaderStatus.ts:1]: Could you change the name of the file to `FileUploadersStatus`?

**Review by richbachman — APPROVED** [review]
Looks good. Just need to resolve the comments from others.
